storage/src/checker.rs

```diff
@@ -91,6 +97,95 @@ impl<S: WritableStorage> NodeStore<Committed, S> {
+    fn split_range_into_leaked_areas(
+        &self,
+        leaked_range: Range<LinearAddress>,
+    ) -> Vec<(LinearAddress, AreaIndex)> {
+        let mut leaked = Vec::new();
+        let mut current_addr = leaked_range.start;
+
+        // First attempt to read the valid stored areas from the leaked range
+        loop {
+            let (area_index, area_size) = match self.find_stored_area_index_and_size(current_addr) {
+                Ok(area_index_and_size) => area_index_and_size,
+                Err(e) => {
+                    warn!("Error reading stored area at {current_addr}: {e}");
+                    break;
+                }
+            };
+
+            let next_addr = current_addr
+                .checked_add(area_size)
+                .expect("address overflow is impossible");
+            match next_addr.cmp(&leaked_range.end) {
+                Ordering::Equal => {
+                    // we have reached the end of the leaked area, done
+                    leaked.push((current_addr, area_index));
+                    return leaked;
+                }
+                Ordering::Greater => {
+                    // the last area extends beyond the leaked area - this means the leaked area overlaps with an area we have visited
+                    warn!(
+                        "Leaked area extends beyond {leaked_range:?}: {current_addr} -> {next_addr}"
+                    );
+                    break;
+                }
+                Ordering::Less => {
+                    // continue to the next area
+                    leaked.push((current_addr, area_index));
+                    current_addr = next_addr;
+                }
+            }
+        }
+
+        // We encountered an error, split the rest of the leaked range into areas using heuristics
+        // The heuristic is to split the leaked range into areas of the largest size possible - we assume `AREA_SIZE` is in ascending order
+        for (area_index, area_size) in AREA_SIZES.iter().enumerate().rev() {
+            loop {
+                let next_addr = current_addr
+                    .checked_add(*area_size)
+                    .expect("address overflow is impossible");
+                if next_addr <= leaked_range.end {
+                    leaked.push((current_addr, area_index as AreaIndex));
+                    current_addr = next_addr;
+                } else {
+                    break;
+                }
+            }
+        }
+
+        // we assume that all areas are aligned to `MIN_AREA_SIZE`, in which case leaked ranges can always be split into free areas perfectly
+        debug_assert!(current_addr == leaked_range.end);
+        leaked
+    }
+
+    // Find the area index and size of the stored area at the given address if the area is valid.
+    // TODO: there should be a way to read stored area directly instead of try reading as a free area then as a node
+    fn find_stored_area_index_and_size(
```

Comment:
nit: naming: read_freed_area

This would have been easier to name if we had a type FreedArea or something, but I think that's too intrusive now.

Reply:
I renamed it to `read_leaked_area` since it may be free or allocated.